presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -264,6 +264,16 @@ public void setNoMorePages()
 
         masterBuffer.setNoMorePages();
 
+        // process any pending reads from the client buffers
+        for (ClientBuffer clientBuffer : safeGetBuffersSnapshot()) {
```

Comment:
I'm not sure this is correct.  The `ClientBuffer` has a push and a pull mode.  The push mode uses `enueuPages` and `setNoMorePages` and the pull mode uses `loadPagesIfNecessary`.  I'm not confident that using `setNoMorePages` in a pull mode would do what we want.  Instead, I suggest we simply call `loadPagesIfNecessary` on every `ClientBuffer` and then modify `loadPagesIfNecessary` to properly handle being called with a `!pagesSupplier.mayHaveMorePages()`.